src/gurobi_optimods/line_optimization.py

```diff
@@ -0,0 +1,404 @@
+        logger.info("column source not present in demand_data")
+        missingData = True
+    if "target" not in demand_data.columns:
+        logger.info("column target not present in demand_data")
+        missingData = True
+    if "demand" not in demand_data.columns:
+        logger.info("column demand not present in demand_data")
+        missingData = True
+    if node_data.isnull().values.any():
+        logger.info("some value is nan or format is not correct for node_data")
+        missingData = True
+    if edge_data.isnull().values.any():
+        logger.info("some value is nan or format is not correct for edge_data")
+        missingData = True
+    if line_data.isnull().values.any():
+        logger.info("some value is nan or format is not correct for line_data")
+        missingData = True
+    if linepath_data.isnull().values.any():
+        logger.info("some value is nan or format is not correct for linepath_data")
+        missingData = True
+    if demand_data.isnull().values.any():
+        logger.info("some value is nan or format is not correct for demand_data")
+        missingData = True
+    if (demand_data < 0).values.any():
+        logger.info(
+            "All demand should be non-negative! Some value is negative in demand_data"
+        )
+        missingData = True
+    if missingData:
+        return -1, []
+
+    # prepare data
+    nodes = node_data.set_index("number").to_dict("index")
+    edges = edge_data.set_index(["source", "target"]).to_dict("index")
+    lines = line_data.set_index("linename").to_dict("index")
+    linepaths = (
+        linepath_data.set_index("linename")
+        .groupby(["linename"])
+        .apply(lambda x: [(k, v) for k, v in zip(x["edgeSource"], x["edgeTarget"])])
+    )
+    demands = demand_data.set_index(["source", "target"]).to_dict()["demand"]
+
+    # remove edges that are not covered by any lines
+    for k in list(edges.keys()):
+        found = False
+        for l in linepaths.index:
+            if k in linepaths[l]:
+                found = True
+                break
+        if found == False:
+            del edges[k]
+
+    if shortestPaths:
+        if nx is None:
+            logger.info(
+                "Networkx is needed for strategy 1 but not available. Using strategy 2 instead."
+            )
+            return allowAllPaths(
+                nodes,
+                edges,
+                lines,
+                linepaths,
+                demands,
+                demand_data,
+                frequencies,
+                create_env,
+            )
+        return allShortestPaths(
+            nodes, edges, edge_data, lines, linepaths, demands, frequencies, create_env
+        )
+    else:
+        return allowAllPaths(
+            nodes,
+            edges,
+            lines,
+            linepaths,
+            demands,
+            demand_data,
+            frequencies,
+            create_env,
+        )
+
+
+# Strategy 1:
```

Comment:
These comments before functions should be docstrings instead.

Reply:
I updated this, hopefully in the correct way.

Reply:
Looks good!